### PR TITLE
Plumbing for realtime consuming segment statistics

### PR DESCRIPTION
**Pull request by mcvsubbu**
In order to allocate memory efficiently for dictionary for a column, we can attempt to use
the statistics gathered from previously completed segments.

Added a class to keep track of statistics for N previously completed segments for a table on local disk
of each server. For now, we have code to store average size (in case of a string column) and cardinality
(for all columns). These can be used when initializing a dictionary for the next segment.

Some methods are synchronized since we expect that segments may be completing at the same time for all
partitions hosted in a machine.

Added a test class.

Added methods in off-heap dictionary classes to get the statistics we want

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+
+/*
+ * Keeps history of statistics for segments consumed from kafka for a single table.
+ */
+public class RealtimeSegmentStatsHistory implements Serializable {
+  private static final long serialVersionUID = 1L;  // Change this if a new field is added to this class
+  private static int MAX_NUM_ENTRIES = 16;  // Max number of past segments for which stats are kept
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java]: I thought only final variables should be all caps. Please check once

Or is there a way to keep this final for safety ?

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    return _entries[index];
+  }
+
+  @Override
+  public String toString() {
+    return "cursor=" + getCursor() + ",numEntries=" + getArraySize() + ",isFull=" + isFull();
+  }
+
+  public synchronized void serializeInto(File outFile) throws IOException {
+    OutputStream os = new FileOutputStream(outFile);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java]: Both streams need to be closed. Prefer try-with-resource or close in finally

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    ObjectOutputStream obos = new ObjectOutputStream(os);
+    obos.writeObject(this);
+    obos.flush();
+    os.close();
+  }
+
+  public static RealtimeSegmentStatsHistory deserialzeFrom(File inFile) throws IOException, ClassNotFoundException {
+    InputStream is = new FileInputStream(inFile);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java]: close both and close in finally or try-with-resource

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java (hunk 1 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java]: I had set it to be final, but wanted to change it for testing reasons. I will add a comment saying that the static should really be final. Not sure if there are other ways of testing this.

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java (hunk 2 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java]: ok

File: pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java (hunk 3 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/realtime/impl/RealtimeSegmentStatsHistory.java]: ok

**Pull request merged**